File: packages/cortex-js/tests/resources/auth.test.ts
import { describe, it, expect } from 'vitest';
import { ValidationError } from '../../src/errors.js';
import { mockFetch, createClient } from '../helpers.js';

describe('auth', () => {
  describe('login', () => {
    it('logs in with email and password', async () => {
      const mockResponse = {
        token: 'jwt-token-123',
        user: { id: 'u-1', email: 'test@example.com', createdAt: '2024-01-01' },
      };
      const fetch = mockFetch(mockResponse);
      const client = createClient(fetch);

      const result = await client.auth.login({ email: 'test@example.com', password: 'secret' });

      expect(result.token).toBe('jwt-token-123');
      expect(result.user.email).toBe('test@example.com');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/auth/login');
      expect(opts.method).toBe('POST');
    });

    it('throws ValidationError when email is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.auth.login({ email: '', password: 'secret' })).rejects.toThrow(ValidationError);
    });

    it('throws ValidationError when password is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.auth.login({ email: 'test@example.com', password: '' })).rejects.toThrow(ValidationError);
    });
  });

  describe('signup', () => {
    it('signs up a new account', async () => {
      const mockResponse = {
        token: 'jwt-token-456',
        user: { id: 'u-2', email: 'new@example.com', createdAt: '2024-01-01' },
      };
      const fetch = mockFetch(mockResponse);
      const client = createClient(fetch);

      const result = await client.auth.signup({ email: 'new@example.com', password: 'secret', name: 'New User' });

      expect(result.token).toBe('jwt-token-456');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/auth/signup');
      expect(opts.method).toBe('POST');
    });

    it('throws ValidationError when email is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.auth.signup({ email: '', password: 'secret' })).rejects.toThrow(ValidationError);
    });

    it('throws ValidationError when password is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.auth.signup({ email: 'test@example.com', password: '' })).rejects.toThrow(ValidationError);
    });
  });

  describe('me', () => {
    it('gets current user profile', async () => {
      const mockResponse = { id: 'u-1', email: 'test@example.com', name: 'Test', createdAt: '2024-01-01' };
      const fetch = mockFetch(mockResponse);
      const client = createClient(fetch);

      const result = await client.auth.me();

      expect(result.email).toBe('test@example.com');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/auth/me');
      expect(opts.method).toBe('GET');
    });
  });

  describe('updateProfile', () => {
    it('updates current user profile', async () => {
      const mockResponse = { id: 'u-1', email: 'test@example.com', name: 'Updated', createdAt: '2024-01-01' };
      const fetch = mockFetch(mockResponse);
      const client = createClient(fetch);

      const result = await client.auth.updateProfile({ name: 'Updated' });

      expect(result.name).toBe('Updated');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/auth/me');
      expect(opts.method).toBe('PATCH');
    });
  });

  describe('changePassword', () => {
    it('changes password', async () => {
      const fetch = mockFetch({ success: true });
      const client = createClient(fetch);

      const result = await client.auth.changePassword({
        currentPassword: 'old',
        newPassword: 'new',
      });

      expect(result.success).toBe(true);

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/auth/change-password');
      expect(opts.method).toBe('POST');
    });

    it('throws ValidationError when currentPassword is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(
        client.auth.changePassword({ currentPassword: '', newPassword: 'new' }),
      ).rejects.toThrow(ValidationError);
    });

    it('throws ValidationError when newPassword is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(
        client.auth.changePassword({ currentPassword: 'old', newPassword: '' }),
      ).rejects.toThrow(ValidationError);
    });
  });
});

File: packages/cortex-dart/test/resources/auth_test.dart
import 'dart:convert';

import 'package:nfinitmonkeys_cortex_sdk/cortex_sdk.dart';
import 'package:http/http.dart' as http;
import 'package:test/test.dart';

import '../helpers.dart';

Map<String, dynamic> sampleAuthToken() => {
      'token': 'jwt-token-abc123',
      'user': {
        'id': 'user-123',
        'email': 'alice@example.com',
        'name': 'Alice',
        'role': 'admin',
      },
    };

Map<String, dynamic> sampleAuthUser() => {
      'id': 'user-123',
      'email': 'alice@example.com',
      'name': 'Alice',
      'role': 'admin',
    };

void main() {
  group('AuthResource', () {
    late CortexClient client;

    tearDown(() => client.close());

    test('login sends correct body', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, sampleAuthToken(), requests),
      );

      final result = await client.auth.login(
        email: 'alice@example.com',
        password: 'password123',
      );

      expect(result.token, 'jwt-token-abc123');
      expect(result.user?.email, 'alice@example.com');

      final body = jsonDecode(requests.first.body) as Map<String, dynamic>;
      expect(body['email'], 'alice@example.com');
      expect(body['password'], 'password123');
    });

    test('login validates empty email', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.auth.login(email: '', password: 'pass'),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('login validates empty password', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.auth.login(email: 'a@b.com', password: ''),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('signup sends correct body', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, sampleAuthToken(), requests),
      );

      final result = await client.auth.signup(
        email: 'alice@example.com',
        password: 'password123',
        name: 'Alice',
      );

      expect(result.token, 'jwt-token-abc123');

      final body = jsonDecode(requests.first.body) as Map<String, dynamic>;
      expect(body['email'], 'alice@example.com');
      expect(body['password'], 'password123');
      expect(body['name'], 'Alice');
    });

    test('signup validates empty email', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.auth.signup(email: '', password: 'pass'),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('signup validates empty password', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.auth.signup(email: 'a@b.com', password: ''),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('me returns current user', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, sampleAuthUser()),
      );

      final user = await client.auth.me();
      expect(user.id, 'user-123');
      expect(user.email, 'alice@example.com');
      expect(user.name, 'Alice');
      expect(user.role, 'admin');
    });

    test('updateProfile sends PATCH request', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, sampleAuthUser(), requests),
      );

      final user = await client.auth.updateProfile(name: 'Alice Updated');

      expect(user.id, 'user-123');
      expect(requests.first.method, 'PATCH');
      expect(requests.first.url.path, contains('/auth/me'));

      final body = jsonDecode(requests.first.body) as Map<String, dynamic>;
      expect(body['name'], 'Alice Updated');
    });

    test('changePassword sends correct body', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, {'ok': true}, requests),
      );

      await client.auth.changePassword(
        currentPassword: 'old-pass',
        newPassword: 'new-pass',
      );

      expect(requests.first.method, 'POST');
      expect(
        requests.first.url.path,
        contains('/auth/change-password'),
      );

      final body = jsonDecode(requests.first.body) as Map<String, dynamic>;
      expect(body['currentPassword'], 'old-pass');
      expect(body['newPassword'], 'new-pass');
    });

    test('changePassword validates empty currentPassword', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.auth.changePassword(
          currentPassword: '',
          newPassword: 'new',
        ),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('changePassword validates empty newPassword', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.auth.changePassword(
          currentPassword: 'old',
          newPassword: '',
        ),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('handles 401 errors', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        maxRetries: 0,
        httpClient: mockClient(401, 'Unauthorized'),
      );

      expect(
        () => client.auth.me(),
        throwsA(isA<CortexAuthenticationException>()),
      );
    });
  });
}
